go/vt/sqlparser/tracked_buffer.go

```diff
@@ -288,14 +314,26 @@ func areBothISExpr(op Expr, val Expr) bool {
 // WriteArg writes a value argument into the buffer along with
 // tracking information for future substitutions.
 func (buf *TrackedBuffer) WriteArg(prefix, arg string) {
+	length := len(prefix) + len(arg)
 	buf.bindLocations = append(buf.bindLocations, BindLocation{
 		Offset: buf.Len(),
-		Length: len(prefix) + len(arg),
+		Length: length,
 	})
+	buf.Grow(length)
 	buf.WriteString(prefix)
 	buf.WriteString(arg)
 }
 
+// WriteInt writes a signed integer into the buffer.
+func (buf *TrackedBuffer) WriteInt(v int64) {
+	buf.WriteString(strconv.FormatInt(v, 10))
+}
+
+// WriteUint writes an unsigned integer into the buffer.
+func (buf *TrackedBuffer) WriteUint(v uint64) {
+	buf.WriteString(strconv.FormatUint(v, 10))
```

Comment:
Note that this is still allocating memory when the small-number optimization doesn't trigger, so that isn't great. I think it may be time to leave behind the `*strings.Builder` used here and instead implementing building ourselves, which would allow us to simplify many things, including using `strconv.AppendInt` instead of the `Format` versions. Maybe for another PR!

Reply:
Yeah, I assume you're thinking something more specialized like what `zap` does? https://github.com/uber-go/zap/blob/master/buffer/buffer.go